src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -823,13 +823,46 @@ trait Infer extends Checkable {
         case _                                                              => onRight
       }
     }
+
     private def isAsSpecificValueType(tpe1: Type, tpe2: Type, undef1: List[Symbol], undef2: List[Symbol]): Boolean = tpe1 match {
       case PolyType(tparams1, rtpe1) =>
         isAsSpecificValueType(rtpe1, tpe2, undef1 ::: tparams1, undef2)
       case _                         =>
         tpe2 match {
           case PolyType(tparams2, rtpe2) => isAsSpecificValueType(tpe1, rtpe2, undef1, undef2 ::: tparams2)
-          case _                         => existentialAbstraction(undef1, tpe1) <:< existentialAbstraction(undef2, tpe2)
+          case _                         =>
+
+            if (settings.YdottySpecificity) {
+              // Backport of fix for https://github.com/scala/bug/issues/2509
+              // from Dotty https://github.com/lampepfl/dotty/commit/89540268e6c49fb92b9ca61249e46bb59981bf5a
+              val flip = new TypeMap(trackVariance = true) {
+                var subst: Map[Symbol, Symbol] = Map.empty
+                def apply(tp: Type): Type = tp match {
+                  case TypeRef(pre, sym, args) if variance > 0 && sym.typeParams.exists(_.isContravariant) =>
+                    val flippedSym =
+                      subst.get(sym) match {
+                        case None =>
+                          val flippedClassSym = sym.owner.newClassSymbol(sym.name.toTypeName, sym.pos, sym.flags)
+                          val flippedParams = sym.rawInfo.typeParams.map { param =>
+                            val flippedFlags = if(param.isContravariant) (param.flags&(~CONTRAVARIANT))|COVARIANT else param.flags
+                            flippedClassSym.newTypeParameter(param.name.toTypeName, param.pos, flippedFlags) setInfo param.info
+                          }
```

Comment:
`param.info` might refer to the original type parameters, so more substitution is needed. Cloning the class upfront takes care of all the rewiring.

```
scala> class C[-A <: C[A]]
defined class C

scala> val CClass = symbolOf[C[_]]
CClass: $r.intp.global.TypeSymbol = class C

scala> val CClass_clone = CClass.cloneSymbol(CClass.owner)
CClass_clone: CClass.TypeOfClonedSymbol = class C

scala> CClass_clone.rawInfo.typeParams.foreach(sym => if (sym.isContravariant) sym.resetFlag(Flag.CONTRAVARIANT).setFlag(Flag.COVARIANT))

scala> CClass_clone.defString
res3: String = class C[+A <: C[A]] extends AnyRef

scala> {settings.uniqid.value = true; println(CClass.defString); println(CClass_clone.defString); settings.uniqid.value = false}
class C#132392[-A#132393 <: C#132392[A#132393]] extends AnyRef#1932
class C#145639[+A#145640 <: C#132392[A#145640]] extends AnyRef#1932
```